src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -67,68 +81,266 @@ where
+    ) -> CopyToS3Uploader {
+        let (bucket, path_prefix) = Self::extract_s3_bucket_path(&sink_connection.prefix);
+        CopyToS3Uploader {
+            desc: sink_connection.desc,
+            sdk_config,
+            format: sink_connection.format,
+            file_name_prefix,
+            bucket,
+            path_prefix,
+            max_file_size: sink_connection.max_file_size,
+            file_index: 0,
+            single_file_uploader: None,
+            buf: Vec::new(),
+        }
+    }
+
+    async fn start_new_file_upload(&mut self) -> Result<(), anyhow::Error> {
+        self.file_index += 1;
+        let file_path = self.file_path();
+        let bucket = self.bucket.clone();
+        info!(
+            "starting upload at bucket: {}, file {}",
+            &bucket, &file_path
+        );
+        let uploader = S3MultiPartUploader::try_new(
+            &self.sdk_config,
+            bucket,
+            file_path,
+            S3MultiPartUploaderConfig {
+                part_size_limit: ByteSize::mib(10).as_u64(),
+                file_size_limit: self.max_file_size,
+            },
+        )
+        .await?;
+        self.single_file_uploader = Some(uploader);
+        Ok(())
+    }
+
+    fn extract_s3_bucket_path(prefix: &str) -> (String, String) {
+        let uri = Uri::from_str(prefix).expect("valid s3 url");
+        let bucket = uri.host().expect("s3 bucket");
+        let path = uri.path().trim_start_matches('/').trim_end_matches('/');
+        (bucket.to_string(), path.to_string())
+    }
+
+    fn file_path(&self) -> String {
+        format!(
+            "{}/{}-{:04}.csv",
+            self.path_prefix, self.file_name_prefix, self.file_index
+        )
+    }
+
+    /// Finishes any remaining in progress upload.
+    async fn finish_ongoing_upload(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(uploader) = self.single_file_uploader.take() {
+            uploader.finish().await?;
+        }
+        Ok(())
+    }
+
+    async fn upload_buffer(&mut self) -> Result<(), anyhow::Error> {
+        match self.single_file_uploader.as_mut() {
+            Some(uploader) => Ok(uploader.add_chunk(&self.buf).await?),
+            None => Err(anyhow!("trying to upload without initializing upload!")),
+        }
+    }
+
+    // Appends the row to the in-progress upload or creates a new upload if it
+    // will exceed the max file size.
+    async fn append_row(&mut self, row: &Row) -> Result<(), anyhow::Error> {
+        // encode the row and write to temp buffer.
+        self.buf.clear();
+        encode_copy_format(self.format.clone(), row, self.desc.typ(), &mut self.buf)
+            .map_err(|_| anyhow!("error encoding row"))?;
+
+        if self.single_file_uploader.is_none() {
+            self.start_new_file_upload().await?;
+        }
+
+        let Some(uploader) = self.single_file_uploader.as_mut() else {
+            unreachable!("this is initialized above");
+        };
```

Comment:
```suggestion
        let uploader = match &mut self.single_file_uploader {
            None => self.start_new_file_upload().await?,
            Some(uploader) => uploader,
        };
```

Reply:
So, I was trying to have the method return the uploader (and I need `&mut uploader`), but I was getting borrow checker errors when I was trying to send `&self.buf` to the uploader's `add_chunk` method.

Reply:
Ah, yeah, you need the partial borrow. Scratch that suggestion, then!